### PR TITLE
add some function to ssl.c for client side session resumption.

### PR DESCRIPTION
**Pull request by dark264sh**
Motivation:
It need getSession, setSession and freeSession for client side session resumption.

Modifications:
Add getSession(SSL_get1_session), setSession(SSL_set_session), freeSession(SSL_SESSION_free) method.

Result:
Be able to handle session for client side session resumption.

### CODE REVIEW

**Comment by Scottmitch** [comment]
Thanks for the PR @dark264sh! Can you sign the [ICLA](https://docs.google.com/spreadsheet/viewform?formkey=dHBjc1YzdWhsZERUQnhlSklsbG1KT1E6MQ) please?

**Comment by dark264sh** [comment]
I've signed the ICLA.

**Comment by normanmaurer** [comment]
LGTM... @Scottmitch WDYT ?

File: openssl-dynamic/src/main/java/org/apache/tomcat/jni/SSL.java [hunk 1]
```diff
@@ -881,4 +881,32 @@ public static native void setCertificateBio(
      * Free x509 chain ({@code STACK_OF(X509)} pointer).
      */
     public static native void freeX509Chain(long x509Chain);
+
+    /**
+     * Call SSL_get1_session.
+     * https://www.openssl.org/docs/man1.1.0/ssl/SSL_get1_session.html
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/java/org/apache/tomcat/jni/SSL.java]: make this a html hyperlink? `<a href="">..</a>`

Comment (Scottmitch) [inline openssl-dynamic/src/main/java/org/apache/tomcat/jni/SSL.java]: same comment for links below

Comment (dark264sh) [inline openssl-dynamic/src/main/java/org/apache/tomcat/jni/SSL.java]: OK.

**Comment by dark264sh** [comment]
Now we are testing. We has a plan to submit a PR for Netty.

**Comment by normanmaurer** [comment]
@dark264sh nice... would be cool to see how this is used with netty together before merge this :)

**Comment by Scottmitch** [comment]
@normanmaurer +1.

We currently hide all of the session stuff behind other APIs today (e.g. `getTime`). I'd be curious if we could do the same thing here. Otherwise I would like to see the lifetime implications these new APIs on the session objects.

**Comment by dark264sh** [comment]
I understand your concern. I have not thought about the lifetime of ssl session. After thinking and studying about ssl session, I will send pull request again. Thank you for your review.

**Comment by Scottmitch** [comment]
sounds good @dark264sh ... feel free to open a new PR when you have a better understanding for lifetime and usage from java.

**Pull request closed without merging**